modules/01_Theory.md

```diff
@@ -0,0 +1,502 @@
+
+### The role of declaration files
+
+In the previous example, we saw the “remapping” part of module resolution working between input and output files. But what happens when we import library code? Even if the library was written in TypeScript, it may not have published its source code. If we can’t rely on mapping the library’s JavaScript files back to a TypeScript file, we can verify that our import works at runtime, but how do we accomplish our second goal of assigning types?
+
+> Why not analyze the JavaScript files directly to get type information? [TODO]
+
+This is where declaration files (`.d.ts`, `.d.mts`, etc.) come into play. The best way to understand how declaration files are interpreted is to understand where they come from. When you run `tsc --declaration` on an input file, you get one output JavaScript file and one output declaration file:
+
+```mermaid
+graph TB
+  main.ts --> main.js
+  main.ts --> main.d.ts
+  style main.js stroke-dasharray: 5 5
+  style main.d.ts stroke-dasharray: 15 4
+```
+
+Because of this relationship, the compiler _assumes_ that wherever it sees a declaration file, there is a corresponding JavaScript file that is perfectly described by the type information in the declaration file. For performance reasons, in every module resolution mode, the compiler always looks for TypeScript and declaration files first, and if it finds one, it doesn’t continue looking for the corresponding JavaScript file—if it finds a TypeScript input file, it knows a JavaScript file _will_ exist after compilation, and if it finds a declaration file, it knows a compilation (perhaps someone else’s) already happened and created a JavaScript file at the same time as the declaration file.
+
+“But wait! Plenty of declaration files are written by hand, _not_ generated by `tsc`. Ever heard of DefinitelyTyped?” you might object. And it’s true—hand-writing declaration files, or even moving/copying/renaming them to represent outputs of an external build tool, is a dangerous, error-prone venture. Luckily, there are only three rules that have to be followed to ensure that hand-placed declaration files don’t mislead the module resolution process:
+
+1. **Every declaration file represents exactly one JavaScript file.** If you try to use a single declaration file to represent two implementation files—an ES module and a CJS module with similar contents, for example—things can go off the rails. On the other hand, writing a declaration file that represents _zero_ JavaScript files is also dangerous, because it allows an import to be written to a JS file that doesn’t exist. Typically, such an import statement will be erased as long as it only imports types, but can be preserved under `verbatimModuleSyntax`, leading to a runtime error.
```

Comment:
How should this be caveated as regards to export maps?

Reply:
Not sure what you mean?